pkg/serviceaccount/util.go

```diff
@@ -79,3 +88,80 @@ func IsServiceAccountToken(secret *v1.Secret, sa *v1.ServiceAccount) bool {
 
 	return true
 }
+
+// PublicJWKSFromKeys constructs a JSONWebKeySet from a list of keys. The key
+// set will only contain the public keys associated with the input keys.
+func PublicJWKSFromKeys(in []interface{}) (*jose.JSONWebKeySet, errors.Aggregate) {
+	// Decode keys into a JWKS.
+	var keys jose.JSONWebKeySet
+	var errs []error
+	for i, key := range in {
+		var pubkey *jose.JSONWebKey
+		var err error
+
+		switch k := key.(type) {
+		case interface {
+			Public() crypto.PublicKey
+		}:
+			// This is a private key. Get its public key
+			pubkey, err = jwkFromPublicKey(k.Public())
+		default:
+			pubkey, err = jwkFromPublicKey(k)
+		}
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error constructing JWK for key #%d: %v", i, err))
+			continue
+		}
+
+		if !pubkey.Valid() {
+			errs = append(errs, fmt.Errorf("key #%d not valid", i))
+			continue
+		}
+		keys.Keys = append(keys.Keys, *pubkey)
+	}
+	if len(errs) != 0 {
+		return nil, errors.NewAggregate(errs)
+	}
+	return &keys, nil
+}
+
+func jwkFromPublicKey(publicKey crypto.PublicKey) (*jose.JSONWebKey, error) {
+	alg, err := algorithmFromPublicKey(publicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	keyID, err := keyIDFromPublicKey(publicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &jose.JSONWebKey{
+		Algorithm: string(alg),
+		Key:       publicKey,
+		KeyID:     keyID,
+		Use:       "sig",
+	}, nil
+}
+
+func algorithmFromPublicKey(publicKey crypto.PublicKey) (jose.SignatureAlgorithm, error) {
+	switch pk := publicKey.(type) {
+	case *rsa.PublicKey:
+		return jose.RS256, nil
```

Comment:
is this always the algorithm for rsa?

Reply:
good catch. I guess we should also expect RS384 and RS512? @mikedanese wdyt?

Reply:
K8s only supports RS256 at present. Added a comment to also update this file if that support is extended.